core/src/main/scala/kafka/log/Log.scala

```diff
@@ -677,6 +672,47 @@ class Log(val dir: File,
   }
 
   /**
+    * Find any segments that match the user-supplied predicate UNLESS it is the final segment
+    * and it is empty (since we would just end up re-creating it)
+    * @param predicate A function that takes in a single log segment and returns true iff it is deletable
+    * @return the segments ready to be deleted
+    */
+  private def deletableSegments(predicate: LogSegment => Boolean) = {
+    val lastEntry = segments.lastEntry
+    if (lastEntry == null) Seq.empty
+    else logSegments.takeWhile(s => predicate(s) && (s.baseOffset != lastEntry.getValue.baseOffset || s.size > 0))
```

Comment:
This takeWhile makes the javadoc slightly incorrect:

> Find any segments that match the user-supplied predicate UNLESS ...

It would be more correct to say something like:

> Find segments starting from the oldest until the the user-supplied predicate is false. A final segment that is empty will never be returned (since we would just end up re-creating it).